taiga_halo2/examples/sudoku/app/valid_sudoku.rs

```diff
@@ -0,0 +1,447 @@
+            layouter.namespace(|| "y"),
+            config.advices[1],
+            Value::known(y),
+        )?;
+
+        //
+        // STEP 2
+        //
+
+        // gamma = Poseidon(x, y)
+        let gamma = {
+            let poseidon_message: [AssignedCell<pallas::Base, pallas::Base>; 2] = [x_cell, y_cell];
+
+            // let poseidon_message = [u[0], u[1]];
+            let poseidon_hasher = halo2_gadgets::poseidon::Hash::<_, _, P128Pow5T3, _, 3, 2>::init(
+                config.poseidon_chip(),
+                layouter.namespace(|| "Poseidon init"),
+            )?;
+            poseidon_hasher.hash(
+                layouter.namespace(|| "Poseidon hash (nk, rho)"),
+                poseidon_message,
+            )?
+        };
+
+        //
+        // STEP 3
+        //
+        // cells for the sudoku entries.
+        let sudoku_cells: Vec<AssignedCell<_, _>> = self
+            .sudoku
+            .concat()
+            .iter()
+            .map(|x| {
+                assign_free_advice(
+                    layouter.namespace(|| "sudoku_cell"),
+                    config.advices[0],
+                    Value::known(pallas::Base::from_u128(*x as u128)),
+                )
+                .unwrap()
+            })
+            .collect();
+
+        // cells for (1..10)
+        let cell_integers: Vec<AssignedCell<_, _>> = (1..10)
+            .map(|i| {
+                assign_free_advice(
+                    layouter.namespace(|| "cell i"),
+                    config.advices[0],
+                    Value::known(pallas::Base::from(i)),
+                )
+                .unwrap()
+            })
+            .collect();
+
+        // rows
+        let rows: Vec<Vec<AssignedCell<Fp, Fp>>> =
+            sudoku_cells.chunks(9).map(|row| row.to_vec()).collect();
+        // cols
+        let cols: Vec<Vec<AssignedCell<Fp, Fp>>> = (1..10)
+            .map(|i| {
+                let col: Vec<AssignedCell<Fp, Fp>> = sudoku_cells
+                    .chunks(9)
+                    .map(|row| row[i - 1].clone())
+                    .collect();
+                col
+            })
+            .collect();
+        // small squares
+        let mut squares: Vec<Vec<AssignedCell<Fp, Fp>>> = vec![];
+        for i in 1..4 {
+            for j in 1..4 {
+                let sub_lines = &rows[(i - 1) * 3..i * 3];
+
+                let square: Vec<&[AssignedCell<Fp, Fp>]> = sub_lines
+                    .iter()
+                    .map(|line| &line[(j - 1) * 3..j * 3])
+                    .collect();
+                squares.push(square.concat());
+            }
+        }
+
+        // check that rows, cols and small squares are permutations of (1..10)
+        for (cpt, perm) in [rows, cols, squares].concat().iter().enumerate() {
+            // lhs = prod_i (i + gamma)
+            // rhs = -prod_i (sigma(i) + gamma)
+            // compute the circuit corresponding to lhs + rhs (should be zero)
```

Comment:
Are we using `gamma` as a random challenge to check the permutation?

Reply:
Yes. See Lemma A.3 [here](https://eprint.iacr.org/archive/2019/953/1656500364.pdf)

Reply:
I'm not sure if we can use in this way here. It seems that `gamma` is deterministic?

Reply:
It is used as a Fiat-Shamir transform with the result of Lemma A.3 where `a_i` are fixed (`(a_i)_i = range(1,10)`) and `b` corresponds to the permutations (the lines, cols and small squares). So `b` is commited in `gamma` and I think it is okay :-)

Reply:
so the `a_i`(or `cell_integers` in the code) should be in constant column, not advice column?

Reply:
`gamma` is deterministic but in order to get the right `gamma`, you need to have a solution for the sudoku